lib/src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -155,6 +224,184 @@ getStudentTransactions student = do
+                       Assignments.desc,
+                       Submissions.signature,
+                       Transactions.grade,
+                       Transactions.time,
+                       Transactions.idx,
+                       BlockTxs.blk_idx
+
+            from       BlockTxs
+            left join  Transactions
+                   on  tx_hash = Transactions.hash
+
+            left join  Submissions
+                   on  Transactions.submission_hash = Submissions.hash
+
+            left join  StudentAssignments
+                   on  StudentAssignments.assignment_hash = Submissions.assignment_hash
+
+            left join  Assignments
+                   on  StudentAssignments.assignment_hash = Assignments.hash
+
+            where      StudentAssignments.student_addr  =  ?
+                  and  time                            >=  ?
+                  and  Transactions.idx                <> -1
+        |]
+
+    -- Returns `PrivateBlock` in normalized format, with metadata.
+    getBlockData :: DBM m => Word32 -> m BlockData
+    getBlockData blockIdx = do
+        (listToMaybe <$> query getBlockDataQuery (Only blockIdx))
+            `assertJust` BlockWithIndexDoesNotExist blockIdx
+      where
+        -- TODO: prune fields that aren't needed.
+        getBlockDataQuery = [q|
+            -- getBlockData
+            select  idx,
+                    hash,
+                    time,
+                    prev_hash,
+                    atg_delta,
+                    mroot,
+                    mtree
+            from    Blocks
+            where   idx = ?
+        |]
+
+getAllNonChainedTransactions :: DBM m => m [PrivateTx]
+getAllNonChainedTransactions = do
+    query getAllNonChainedTransactionsQuery ()
+  where
+    getAllNonChainedTransactionsQuery = [q|
+        -- getAllNonChainedTransactions
+        select     Submissions.student_addr,
+                   Submissions.contents_hash,
+                   Assignments.course_id,
+                   Assignments.contents_hash,
+                   Assignments.type,
+                   Assignments.desc,
+                   Submissions.signature,
+                   grade,
+                   time
+
+        from       Transactions
+
+        left join  Submissions
+               on  submission_hash = Submissions.hash
+
+        left join  Assignments
+               on  assignment_hash = Assignments.hash
+
+        where  idx = -1
+    |]
+
+getLastBlockIdAndIdx :: DBM m => m (Hash PrivateBlockHeader, Word32)
+getLastBlockIdAndIdx = do
+    fromMaybe (genesisHeaderHash, 1) . listToMaybe <$> query [q|
+        -- getLastBlockIdAndIdx
+        select    hash,
+                  idx
+        from      Blocks
+        order by  idx desc
+        limit     1
+    |] ()
+
+createBlock :: DBM m => Maybe ATGDelta -> m ()
+createBlock delta = do
+    (prev, idx) <- getLastBlockIdAndIdx
+    txs         <- getAllNonChainedTransactions
+
+    transaction $ do
```

Comment:
I guess you can execute some of the `let` bindings strictly outside of the `transaction`, but performance is not a problem now.

Reply:
The `let` binding is almost no-op, it doesn't execute - it just binds. Wherever you declare `foo`, it will be evaluated on the first place it is called.

And I propose optimising after we have benchmarks.

Reply:
... up to the first constructor. We need benchmarks - this might be non-bottleneck.

Reply:
And how does it improve performance? The evaluation has to be done no matter when.